src/detect-ike-key-exchange-payload.c

```diff
@@ -0,0 +1,226 @@
+#define KEYWORD_DOC_SERVER_KEY_EXCHANGE  "ike-keywords.html#ike-server_key_exchange_payload";
+#define BUFFER_NAME_SERVER_KEY_EXCHANGE  "ikev1.server_key_exchange_payload"
+#define BUFFER_DESC_SERVER_KEY_EXCHANGE  "ikev1 server key_exchange payload"
+
+static int g_buffer_client_key_exchange_id = 0;
+static int g_buffer_server_key_exchange_id = 0;
+
+static int DetectClientKeyExchangeSetup(DetectEngineCtx *de_ctx, Signature *s, const char *str)
+{
+    if (DetectBufferSetActiveList(s, g_buffer_client_key_exchange_id) < 0)
+        return -1;
+
+    if (DetectSignatureSetAppProto(s, ALPROTO_IKE) < 0)
+        return -1;
+
+    return 0;
+}
+
+static int DetectServerKeyExchangeSetup(DetectEngineCtx *de_ctx, Signature *s, const char *str)
+{
+    if (DetectBufferSetActiveList(s, g_buffer_server_key_exchange_id) < 0)
+        return -1;
+
+    if (DetectSignatureSetAppProto(s, ALPROTO_IKE) < 0)
+        return -1;
+
+    return 0;
+}
+
+static InspectionBuffer *GetClientKeyExchangeData(DetectEngineThreadCtx *det_ctx,
+        const DetectEngineTransforms *transforms, Flow *_f,
+        const uint8_t _flow_flags, void *txv, const int list_id)
+{
+    InspectionBuffer *buffer = InspectionBufferGet(det_ctx, list_id);
+    if (buffer->inspect == NULL) {
+        const uint8_t *b = NULL;
+        uint32_t b_len = 0;
+
+        if (rs_ike_state_get_key_exchange(txv, &b, &b_len) != 1)
+            return NULL;
+        if (b == NULL || b_len == 0)
+            return NULL;
+
+        InspectionBufferSetup(buffer, b, b_len);
+        InspectionBufferApplyTransforms(buffer, transforms);
+    }
+
+    return buffer;
+}
+
+static InspectionBuffer *GetServerKeyExchangeData(DetectEngineThreadCtx *det_ctx,
+                                          const DetectEngineTransforms *transforms, Flow *_f,
+                                          const uint8_t _flow_flags, void *txv, const int list_id)
+{
+    InspectionBuffer *buffer = InspectionBufferGet(det_ctx, list_id);
+    if (buffer->inspect == NULL) {
+        const uint8_t *b = NULL;
+        uint32_t b_len = 0;
+
+        if (rs_ike_state_get_key_exchange(txv, &b, &b_len) != 1)
+            return NULL;
+        if (b == NULL || b_len == 0)
+            return NULL;
+
+        InspectionBufferSetup(buffer, b, b_len);
+        InspectionBufferApplyTransforms(buffer, transforms);
+    }
+
+    return buffer;
+}
+
+static bool DetectClientKeyExchangeHexadecimalValidateCallback(const Signature *s,
+                                                               const char **sigerror)
+{
+    const SigMatch *sm = s->init_data->smlists[g_buffer_client_key_exchange_id];
+    for ( ; sm != NULL; sm = sm->next)
+    {
+        if (sm->type != DETECT_CONTENT)
+            continue;
+
+        const DetectContentData *cd = (DetectContentData *)sm->ctx;
+
+        for (size_t i = 0; i < cd->content_len; ++i)
+        {
+            if(!isxdigit(cd->content[i]))
```

Comment:
Why is only hexadecimal expected here ?

Reply:
You are right pcre is possible as well, i'll add the reserved chars and rename the test.

Reply:
But why not any character from 0 to 255 ?

Reply:
yes....sounds simple, ok

Reply:
checked DetectContentData->content, it's type is *uint8_t
It should not be possible to be < 0 or > 255. (got a compiler warning as well that this cannot happen).

Better remove this test? Alternative would be constraint hex + pcre-chars .....

Reply:
> Alternative would be constraint hex + pcre-chars 

I am trying to figure out why you want this restriction.
Why could not there be `///` in the content for ikev1 exchange payload instance ?

Reply:
The exchange payload is stored in rust as a hex-encoded string. Afaik exact match with content: and "fuzzy" match with pcre: are supported by sticky buffers.
Just wanted to restrict the rule-content to values which are possible. Or does this make no sense?

Reply:
> The exchange payload is stored in rust as a hex-encoded string

Thanks, this is the interesting point I was missing.

So now, why is this stored as a hex-encoded string ?
We could use the raw data, and avoid to have this content validation function.

Reply:
The key-exchange-payload is the public Diffie-Hellman key, which is just a "byte blob", there is no encoding defined in the protocol. I chose hex-encoding because it's the same encoding used in wireshark, making dev/validation easier for me.

Also i found it nicer to have the eve.json output "compatible" with the rules....

Reply:
....checked other protocols/rules, i'll store nonce/payload as byte-array and will transform to hex just for the logger.

Didn't know that |xx| is supported in the rule-syntax for writing "arbitrary byte matches".

first parser i'm implementing completely myself here, feeling a bit noobish from time to time.....

Reply:
You are doing great :-)

Reply:
thx :)